app/src/main/java/ai/kun/opentrace/worker/BLEClient.kt

```diff
@@ -0,0 +1,142 @@
+    private val INTERVAL_KEY = "interval"
+    private val CLIENT_REQUEST_CODE = 11
+
+    private var mScanning = false
+    private var mConnected = false
+
+    override fun onReceive(context: Context, intent: Intent) {
+        val interval = intent.getIntExtra(INTERVAL_KEY, Constants.BACKGROUND_TRACE_INTERVAL)
+        val pm = context.getSystemService(Context.POWER_SERVICE) as PowerManager
+        val wl = pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, WAKELOCK_TAG)
+        wl.acquire(interval.toLong())
+        synchronized(BLETrace) {
+            // Chain the next alarm...
+            enable(interval)
+            startScan()
+        }
+        wl.release()
+    }
+
+    private fun getAdapter(context: Context): BluetoothAdapter? {
+        return (context.getSystemService(BLUETOOTH_SERVICE) as BluetoothManager).adapter
+    }
+
+    fun enable(interval: Int) {
+        BLETrace.alarmManager.setExactAndAllowWhileIdle(
+            AlarmManager.RTC_WAKEUP,
+            ((System.currentTimeMillis() / interval) * interval) + interval,
+            getPendingIntent(interval)
+        )
+    }
+
+    fun disable(interval: Int) {
+        synchronized(BLETrace) {
+            BLETrace.alarmManager.cancel(getPendingIntent(interval))
+            stopScan()
+        }
+    }
+
+    private fun getPendingIntent(interval: Int): PendingIntent {
+        val intent = Intent(BLETrace.context, BLEClient::class.java)
+        intent.putExtra(INTERVAL_KEY, interval)
+        return PendingIntent.getBroadcast(
+            BLETrace.context,
+            CLIENT_REQUEST_CODE,
+            intent,
+            PendingIntent.FLAG_UPDATE_CURRENT
+        )
+    }
+
+
+    // Scanning
+    private fun startScan() {
+        if (mScanning) {
+            Log.w(TAG,"Already scanning")
+            return
+        }
+
+        val scanFilter = ScanFilter.Builder()
+            .setManufacturerData(
+                Constants.MANUFACTURE_ID,
+                MANUFACTURE_SUBSTRING.toByteArray(StandardCharsets.UTF_8))
+            .build()
+
+        val settings = ScanSettings.Builder()
+            .setScanMode(ScanSettings.SCAN_MODE_BALANCED)
+            .build()
+
+        BLETrace.bluetoothLeScanner.startScan(listOf(scanFilter), settings, BtleScanCallback)
+        BtleScanCallback.handler.postDelayed(Runnable { stopScan() }, SCAN_PERIOD)
+        mScanning = true
+        Log.d(TAG, "+++++++Started scanning.")
+    }
+
+    private fun stopScan() {
+
+        synchronized(BLETrace) {
+            if (mScanning && BLETrace.bluetoothManager.adapter.isEnabled) {
+                BLETrace.bluetoothLeScanner.stopScan(BtleScanCallback)
+                scanComplete()
+            }
+            mScanning = false
+        }
+        Log.d(TAG, "-------Stopped scanning.")
+    }
+
+    private fun scanComplete() {
+        if (BtleScanCallback.mScanResults.isEmpty()) {
+            return
+        }
```

Comment:
This seems to be not needed

Reply:
As I said in the commit comment there is a ton of clean-up I still need to do on this.  I've done a bunch from where I started, but I wasn't sure what I would need from the test app I created.  I still need to clean all that stuff out, but I want to get something we can demo quickly.